Add ffbs-debugbathosts

SmithChart opened the pull request:
This package provides `/etc/bat-hosts`.
his file can be used to map mac-addresses to names in `batctl`.

This file is empty by default and mainly serves as a reminder on how this can be used on the devices itself.

maurerle reviewed (commented):
I would like this package to have a way to be set from the site.conf and overwrite this file in the upgrade process?

Very nice, i did not kbow about this behaviour before :)

SmithChart commented:
> I would like this package to have a way to be set from the site.conf and overwrite this file in the upgrade process?

We have thought about a way to fill in the file during build process with information from e.g. meshviewer / yannic.
But overriding / appending data from `site.conf` during upgrade sounds like a good solution.

maurerle commented:
It would also be possible to have a small script which you call as:
`download_bat_hosts https://map.aachen.freifunk.net/data/meshviewer.json` which inserts all hosts into /etc/bat-hosts on demand, as this is anyway only useful if it is used to analyze something :)

That way, the site.conf config is not spoiled with many entries

T-X commented:
Would love to have an automatic feature to populate /etc/bat-hosts (symlinked from /tmp/bat-hosts, to avoid flash writes). This would make debugging issues so much easier! Another, additional idea would be to populate at least from direct neighbors (via respondd?) as a solution that would work offline as well.

(As is, I don't see much value, it's something that is already documented in the batctl README. But of course wouldn't break anything either, so no reason to not merge this PR as is in the community repository.)

PS: spelling: contian -> contain

maurerle commented:
I just created a small script which does about this:

https://gist.github.com/maurerle/3586f2b8d825a69df0f523f60b731efa

One can install it using:

```
wget https://gist.githubusercontent.com/maurerle/3586f2b8d825a69df0f523f60b731efa/raw/23b68babcc73d2054d3cb0df95e9b6f9d239abcd/update-bat-hosts.lua
chmod +x update-bat-hosts.lua
```

One can then call it using you communities nodes.json:
`update-bat-hosts "https://map.aachen.freifunk.net/data/nodes.json" `

Though this takes quite a few seconds on router cpus but is decent on x86.

Few notes:
* batctl hostnames have to be unique so I added the iface and an index to the name
* batctl has a host length limit of 50 - adding iface and idx and other chars, i used 30 real chars as maximum
* batctl gives warnings if hosts are duplicated, so duplicates are removed (as some people just name their device "Freifunk")

This is available under MIT license and can be integrated into this package if wanted :)

maurerle commented on ffbs-debugbathosts/Makefile:
```diff
@@ -0,0 +1,23 @@
+define Package/$(PKG_NAME)/description
+    This package provides an empty /etc/bat-hosts file.
+    It simply serves as a reminder, that this file exists and how
+    it can be used.
```
We should adjust this :)

SmithChart replied:
Updated to match README and package :)

SmithChart commented:
@maurerle your first review comments came in before my CI/CT had time for a turnaround :-D

Let me explain what I've changed:
I've integrated your script. Thanks for sharing! I have changed a few details:

* The output is written to `/etc/bat-hosts`  instead of `/tmp` and symlinking. A usual flash IC in a router should be capable of handling in the ballpark of 100k writes. Even with 20 writes a day for a year straight you do not exhaust 10% of that. And that does not take into account wear-leveling done by `jffs2`. This is plenty of headroom for a script an admin runs by hand. So no need to take extra rounds here.
* I've added a command line switch for the *gateway-only* mode. And it's off by default. At least for Freifunk Braunschweig this mode does not make sense.

SmithChart commented:
> PS: spelling: contian -> contain

Fixed.

SmithChart commented:
I've also thought about having the file populated at compile time. I would not favor such solution since it would make the build less reproducible: A changed `nodes.json` would affect the binary output.

grische commented on ffbs-debugbathosts/files/usr/sbin/update-bat-hosts:
```diff
@@ -0,0 +1,95 @@
+#!/usr/bin/lua
```
Might be nice to have this in `ffbs-debugbathosts/luasrc` instead of `ffbs-debugbathosts/files` so that the Lua compressor can reduce the size of the script.

grische commented on ffbs-debugbathosts/files/usr/sbin/update-bat-hosts:
```diff
@@ -0,0 +1,95 @@
+local url = arg[1]
+
+local gateway_only = false
+if #arg == 2 and arg[2] == '--gw' then
```
I propose to rename this to `--gw-only`. If I see the parameter `--gw` I would think it includes gws 🙂

grische reviewed (commented):
I know, I'm a bit late to the party, but I have some minor suggestions.

SmithChart commented:
> I know, I'm a bit late to the party, but I have some minor suggestions.

Thanks for the feedback. I'll pack those changes into a new PR in a few days.
